### PR TITLE
Serialize function signature simplifications

### PR DESCRIPTION
**Pull request by metesynnada**
## Which issue does this PR close?

Closes #.

## Rationale for this change

> In general, issuing a blocking call or performing a lot of compute in a future without yielding is problematic, as it may prevent the executor from driving other futures forward.

~~https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html~~

~~Previously, the serialization process was always asynchronous, even if there were no asynchronous calls involved, which was a CPU-bound operation. This is an anti-pattern.~~

This PR simplifies function signatures in `BatchSerializer`.

## What changes are included in this PR?

Make `BatchSerializer` sync ~~, and use it with `tokio::task::spawn_blocking`~~.

## Are these changes tested?

With existing tests.

## Are there any user-facing changes?

No.

### CODE REVIEW

**Comment by alamb** [comment]
>> In general, issuing a blocking call or performing a lot of compute in a future without yielding is problematic, as it may prevent the executor from driving other futures forward.

I think this part of the tokio docs is confusing and somewhat misleading. Specifically what "a lot of compute" means is very dependent on the application.

In this case, I think the actual serialization is done for a `RecordBatch` and then `await` is called to potentially yield control. I don't think serializing a single `RecordBatch` qualifies as "a lot of compute"

I wrote up a detailed justification for using tokio for CPU bound tasks in the following blog post, which I think is still very relevant https://thenewstack.io/using-rustlangs-async-tokio-runtime-for-cpu-bound-tasks/

> Previously, the serialization process was always asynchronous, even if there were no asynchronous calls involved, which was a CPU-bound operation. 

Almost all of a DataFusion plan's execution is CPU bound and not asynchronous, yet they are executed using `async` heavily. I don't think this is a problem (for reasons explained in the blog) and I don't see any reason we would want to treat writing batches differently.

So all in all, I don't agree with the stated rationale of this PR that we should not use the same executor for CPU bounds tasks

**Comment by ozankabak** [comment]
@alamb, would the current approach work well in a setting with a small number of threads and large batch sizes? The scenario that came up today when we discussed this internally was running a deep query (i.e. with a lot of operators) in such a setting. We were wondering if the pull mechanism would get disrupted if serializations would take too long in such an environment.

Maybe the answer is no (and we should measure to see what happens), but I'd love to hear your thought process on such a scenario.

**Comment by alamb** [comment]
> @alamb, would the current approach work well in a setting with a small number of threads and large batch sizes? The scenario that came up today when we discussed this internally was running a deep query (i.e. with a lot of operators) in such a setting. We were wondering if the pull mechanism would get disrupted if serializations would take too long in such an environment.

This likely depends on what "large batch size" means and what type of disruption you are talking about. 

My internal mental model is that if you give the query plan `N` threads, it should be able to keep all N threads busy all of the time. Each output serializer will pull the next batch to serialize (the same thread will likely compute the input to serialize if each of the input futures is `ready`) and then serialize the batch. The CPU should be kept busy the entire time making input or serializing output and there won't be any stalls and the query won't suffer context switching overhead (due to more threads than CPUs)

The larger the `target_batch_size` is 
1. the greater the intermediate memory required (as each intermediate batch takes more)
2. the greater the latency may be between generating subsequent batches of output
3. the lower the CPU overhead is (as there are many things done "per batch" which gets amortized over many more rows)

So I think the application needs to figure out how it wants to trade off these features

> Maybe the answer is no (and we should test/measure to see what happens), but I'd love to hear your thought process on such a scenario.

I am always a fan of measuring / testing.

**Comment by ozankabak** [comment]
Thanks, I will ask @metesynnada to run some experiments so we all can learn if there are any intricacies at play here.

We will report the results here and close this PR if this (1) doesn't help in any scenario, or (2) hurts in some scenario.

**Comment by alamb** [comment]
Marking this as draft to signify there is ongoing work -- please feel free to mark it ready for review if I got that wrong

**Comment by ozankabak** [comment]
Interesting results. @alamb, can you take a look at the measurement script? Let's make sure there are no methodological mistakes before arriving at any conclusions.

**Comment by alamb** [comment]
TLDR is that I don't think this benchmark keeps the number of cores constant 

When I run the first benchmark run that is supposed to use 1 core:

```
Benchmarking parameter group/sink_bs100_tp1/1: Collecting 100 samples in estimated 105.75 s (100 iterations)
...
```

Main uses a single core as epxected
Screenshot 2024-01-10 at 3 47 57 PM

When I run on the `synnada-ai:upstream/spawn-blocking-for-se` branch, the first benchmark uses almost three cores: 
Screenshot 2024-01-10 at 3 40 39 PM

What I think is happening is that the `spawn_blocking` call, as one might expect, runs the workload on a dedicated thread that is not controlled by the tokio runtime's worker thread count

```rust
                    let rt = Builder::new_multi_thread()
                        .worker_threads(worker_thread)
                        .build()
                        .unwrap();
```

The fact that the branch is using 3 threads and main can only use one I think accounts for the speed differences


## Methodology
```shell
docker run -v "/Users/andrewlamb/Downloads/tpch_sf0.1":/data -it --rm ghcr.io/databloom-ai/tpch-docker:main -vf -s 0.1
```

I put the benchmark code on this branch https://github.com/alamb/arrow-datafusion/tree/alamb/write_threads_bench

And ran 
```shell
cargo bench --bench write_threads
```

**Comment by ozankabak** [comment]
Good catch. Tokio offers a `max_blocking_threads` API, but I don't immediately see a way to use that to fix the benchmark. Apparently there used to be a `max_threads` API, which was limiting the total number of threads (async workers + blocking), which would be what we want in this context. However, that doesn't seem to be available anymore. I will consult with @metesynnada tomorrow to see if there is anything we can do to fix the methodology.

**Comment by metesynnada** [comment]
I run the benchmarks in single-core and there is no substantial difference. Thus, I will convert this to spawn again.

**Comment by ozankabak** [comment]
Agreed. I changed the PR title and description to match what this does now -- simplifying function signatures only (no change to `async` vs. `spawn_blocking` behaviors). It looks good to me

**Review by alamb — APPROVED** [review]
LGTM -- thank you.

cc @devinjdangelo
